content/updates/understanding-the-terraform-check-block-feature.md

```diff
@@ -0,0 +1,207 @@
+
+Let's consider an example from Masterpoint's module [masterpointio/terraform-aws-tailscale](https://github.com/masterpointio/terraform-aws-tailscale/tree/feature/check_auth_key) where we validate the desired object state using a data source. The module provisions an AWS EC2 instance that serves as a [Tailscale subnet router](https://tailscale.com/kb/1019/subnets/) device. The subnet router setup is deployed via `userdata`, and upon authentication with a [Tailnet key](https://registry.terraform.io/providers/tailscale/tailscale/latest/docs/resources/tailnet_key) — managed by Terraform — the device is automatically tagged. Correct tagging of the Tailscale subnet router is critical for its proper functioning. We can introduce the checks in [checks.tf](https://github.com/masterpointio/terraform-aws-tailscale/blob/feature/check_auth_key/checks.tf) to validate if the tags were successfully applied.
+
+The first assertions verifies that the device has been tagged; the second one confirms that the tags applied are as expected:
+
+```hcl
+variable "tailnet_name" {
+  type        = string
+  description = <<EOF
+  This unique name of the Tailnet that is used when registering DNS entries, e.g. 'cat-crocodile.ts.net'.
+  See https://tailscale.com/kb/1217/tailnet-name/ for more information.
+  EOF
+}
+
+check "device" {
+  data "tailscale_device" "default" {
+    name     = format("%s.%s", module.this.id, var.tailnet_name)
+    wait_for = "30s"
+  }
+
+  assert {
+    condition     = length(data.tailscale_device.default.tags) > 0
+    error_message = "Device ${data.tailscale_device.default.name} is not tagged."
+  }
+
+  assert {
+    condition     = sort(data.tailscale_device.default.tags) == sort(tolist(local.tailscale_tags))
+    error_message = <<EOF
+    Device ${data.tailscale_device.default.name} is not tagged with the correct tags.
+    The list of expected tags is: [${join(",", local.tailscale_tags)}]
+    EOF
+  }
+}
+```
+
+Example of the failed assertion:
+```sh
+Warning: Check block assertion failed
+  on ../../../terraform-aws-tailscale/checks.tf line 19, in check "device":
+  19:     condition     =  sort(data.tailscale_device.default.tags) == sort(tolist(local.tailscale_tags))
+    ├────────────────
+    │ data.tailscale_device.default.tags is set of string with 1 element
+    │ local.tailscale_tags is tuple with 1 element
+Device mp-automation-tailscale-subnet-router.cat-crocodile.ts.net is not tagged with the correct tags.
+```
+
+### Limitations to consider
```

Comment:
Should this be a Heading 2? 

```suggestion
## Limitations to consider
```

Reply:
> Should this be a Heading 2?

Not really. These limitations are for data sources only. We can merge them into one common block if you feel like it would be better for the reader's perception.

Reply:
Merged this with a more general paragraph about potential pitfalls.